Disable remove unnecessary braces diagnotics for self imports

AmrDeveloper opened the pull request:
Disable `remove unnecessary braces` diagnostic if the there is a `self` inside the bracketed `use`

Fix #15191

lowr commented:
`fn remove_braces()` has some handling of `self` imports. Can you tweak it now that it's obsolete?

AmrDeveloper commented:
> `fn remove_braces()` has some handling of `self` imports. Can you tweak it now that it's obsolete?

Thank you @lowr i removed it and updated the code

lowr commented on crates/ide-diagnostics/src/handlers/useless_braces.rs:
```diff
@@ -15,20 +15,23 @@ pub(crate) fn useless_braces(
 ) -> Option<()> {
     let use_tree_list = ast::UseTreeList::cast(node.clone())?;
     if let Some((single_use_tree,)) = use_tree_list.use_trees().collect_tuple() {
+        // If there is a `self` inside the bracketed `use`, don't show diagnostic.
+        if single_use_tree.syntax().first_token().unwrap().kind() == SyntaxKind::SELF_KW {
+            return Some(());
+        }
```
One more thing: `use a::{b::self};` bypasses this check and we'd provide a quickfix that turn it into `use a::b::self;`, which is invalid. Something like [this line](https://github.com/rust-lang/rust-analyzer/pull/15216/files#diff-0cd524758b7692ad400e6cf58628a1016cbd79bcad5e0e06cfdeb50dd7286356L53) from the removed `fn remove_braces()` should handle this kind of cases well.

AmrDeveloper replied:
is `use a::{b::self};` a valid syntax? because i got a syntax error

`Syntax Error: The self keyword is only allowed as the first segment of a path` 

when i created a test case

```rust
use a::{b::self};

mod a {  pub mod b { } }
```

Is there is a think that i am missing?

lowr replied:
Yeah I've noticed that too. Turns out that's rust-analyzer's bug, because rustc accepts it ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=74dd7261a8a50716049ca4aa9d73346f)).

AmrDeveloper replied:
We can cover it by disabling `syntax-error` and using `check_diagnostics_with_config`

lowr commented:
Thanks!

@bors r+

bors commented:
:pushpin: Commit 54e897368d88650b2990aa62ba39dfee0a4f970c has been approved by `lowr`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust-analyzer) for this repository.

<!-- @bors r=lowr 54e897368d88650b2990aa62ba39dfee0a4f970c -->
<!-- homu: {"type":"Approved","sha":"54e897368d88650b2990aa62ba39dfee0a4f970c","approver":"lowr","queue":"https://bors.rust-lang.org/homu/queue/rust-analyzer"} -->

bors commented:
:hourglass: Testing commit 54e897368d88650b2990aa62ba39dfee0a4f970c with merge 3ea31e0ccfe20685b0272923cb9437589e5826e8...
<!-- homu: {"type":"BuildStarted","head_sha":"54e897368d88650b2990aa62ba39dfee0a4f970c","merge_sha":"3ea31e0ccfe20685b0272923cb9437589e5826e8"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/rust-analyzer/actions/runs/5466811205/jobs/9952574449)
Approved by: lowr
Pushing 3ea31e0ccfe20685b0272923cb9437589e5826e8 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"lowr","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/rust-analyzer/actions/runs/5466811205/jobs/9952574449"},"merge_sha":"3ea31e0ccfe20685b0272923cb9437589e5826e8"} -->
